src/Container/Container.php

```diff
@@ -36,76 +41,121 @@ class Container implements ContainerInterface
      * @param mixed  $value
      * @return Container
      */
-    public function set($key, $value)
+    public function set(string $key, $value): self
     {
+        if (!(\is_string($value) || \is_object($value))) {
+            throw new \InvalidArgumentException('Value must be a class name, an object instance, or a callable');
+        }
+
         $this->dicValues[$key] = $value;
         unset($this->dicObjects[$key]); // In case an object instance was stored for sharing
 
         return $this;
     }
 
+    public function alias(string $alias, string $key): self
+    {
+        $this->aliases[$alias] = $key;
+
+        return $this;
+    }
+
     /**
      * Check if a DIC value/object exists.
      *
      * @param string $key
      * @return bool
      */
-    public function has($key)
+    public function has($key): bool
     {
+        if (array_key_exists($key, $this->aliases)) {
+            $key = $this->aliases[$key];
+        }
+
         return array_key_exists($key, $this->dicValues);
     }
 
     /**
      * @param string $key
      * @return bool
      */
-    public function hasInstance($key)
+    public function hasInstance($key): bool
     {
+        if (array_key_exists($key, $this->aliases)) {
+            $key = $this->aliases[$key];
+        }
+
         return isset($this->dicObjects[$key]);
     }
 
     /**
-     * Get the shared instance of a DIC object, or a DIC value if it's not an object.
+     * Get the shared instance of a DIC object
      *
      * @param string $key
      * @return mixed
      */
     public function get($key)
     {
-        if (!$this->has($key)) {
-            throw new NotFoundException(sprintf('No application value with key "%s" is defined.', $key));
+        if (array_key_exists($key, $this->aliases)) {
+            $key = $this->aliases[$key];
         }
 
         // Get already instantiated object if it exist
         if (isset($this->dicObjects[$key])) {
             return $this->dicObjects[$key];
         }
 
-        // Check if it's an invokable (closure/anonymous function)
-        if (is_object($this->dicValues[$key]) && method_exists($this->dicValues[$key], '__invoke')) {
-            $this->dicObjects[$key] = $this->dicValues[$key]($this);
-
-            return $this->dicObjects[$key];
+        try {
+            if (array_key_exists($key, $this->dicValues)) {
+                if (\is_object($this->dicValues[$key])) {
```

Comment:
To check this on every item in the `dicValues` on every call to `$app->get()` looks a bit ineffective... This is why we need to benchmark this before merging.

One way might be to use separate arrays to store invokable objects instead. That might speed things up a bit :-)

Reply:
I'm not entirely sure what you mean by "every item in the `dicValues`" here.

The only real functional difference from the old code is the call to `resolveInstance()` when the value is a string (an interface has been mapped to a concrete class), or when the key does not exist and should itself be treated as a the name of the class to instantiate.